util/collate/collate.go

```diff
@@ -193,6 +193,21 @@ func CollationName2ID(name string) int {
 	return mysql.DefaultCollationID
 }
 
+// SubstituteMissingCollationToDefault will switch to the default collation if
+// new collations are enabled and the specified collation is not supported.
+func SubstituteMissingCollationToDefault(co string) string {
```

Comment:
```
		cht, coll, err := charset.GetCharsetInfo(val)
		if err != nil {
			logutil.BgLogger().Warn(err.Error())
			cht, coll = charset.GetDefaultCharsetAndCollate()
		}
```

I think this part is missing in the new code. I wonder, can we move the two default check into the validation function, `checkCharacterSet`, and `checkCollation`?

Reply:
You are referring to the check for `case CharacterSetConnection, CharacterSetClient, CharacterSetResults, CharacterSetServer, CharsetDatabase, CharacterSetFilesystem`?

As I understood it looking through the code, there were previously two types of validation that were performed:
- One when calling the validation function
- One when calling the Set function

Calling only the Set function and skipping the validate function is not intended, but this is what I discovered some of the code was doing (so I updated the comment for the set function).

I removed the validation which was occuring when only calling the Set function, since this part is actually a duplicate of calling the validation function (it's existing code, not visible in this PR):
https://github.com/morgo/tidb/blob/becbba1416ec590b8bfcad5523ca85ae873fe5aa/sessionctx/variable/varsutil.go#L120-L129

I didn't find an example where the charset (as only the charset without collation) was calling set without validation, but there are two examples where setting collation bypasses validation. These are the ones which now call  `SubstituteMissingCollationToDefault` prior.

Reply:
> You are referring to the check for case CharacterSetConnection, CharacterSetClient, CharacterSetResults, CharacterSetServer, CharsetDatabase, CharacterSetFilesystem

Yes. It will fallback to `cht, coll = charset.GetDefaultCharsetAndCollate()` if it is an invalid charset. The fallback, however, is missing in the `checkCharacterSet()` validation. Same to the collation validation function, which does not really fallback to `UTF8MB4`.

Now the collation fallback is done by the new `SubstituteMissingCollationToDefault`, which is called twice. Charset fallback is missing. The original code will fallback charset&collation whenever `SetSystemVar` is called. I don't think the logic is equivalent.

So I think it is not 100% duplicated to the validation functions. In my understanding, we should fallback in the validation functions. And remove `SubstituteMissingCollationToDefault` in the future.

Reply:
> The fallback, however, is missing in the checkCharacterSet() validation. Same to the collation validation function, which does not really fallback to UTF8MB4.

This is because the tests will fail if the fallback is added into the validation function:
- In the context of calling `SET` or `SET NAMES` an incorrect value is expected to return an error, not fallback.
- In the context of protocol negotiation (which calls `SetSystemVar` directly and skips the validation function) it is expected to fallback.

This is why I was refering to "two types of validation" being performed, since yes validation functions can substitute values, but in this case it does not know the origin of the request. The solution I have gone with, is that if a fallback is desired, it should be called before (or in this case instead of) the validation function.

For collations, there is `SubstituteMissingCollationToDefault`, but to as to why there is no corresponding function for Charset:

```
morgo@ubuntu:~/go/src/github.com/morgo/tidb$ grep -RF SetSystemVar . | grep 'Char'  | grep -vF '_test'
./executor/simple.go:	err := sessionVars.SetSystemVar(variable.CharsetDatabase, dbinfo.Charset)
grep: ./bin/tidb-server: binary file matches
./executor/set.go:	err = sessionVars.SetSystemVar(variable.CharacterSetConnection, csDb)
<snip>
```

I tested `executor/set.go` manually with "SET names boguscharset" (I will also add a test for it \*). It still worked as expected because the code just before it calls this does validate the charset exists. It needs to do this to validate collation + charset together, and return the specific "mismatch" error (i.e. relying only on the validate function means we can't test the interaction of two variables together). I am going to push a change however, and make the `SET NAMES` calls in `executor/set.go` also call the validation function\*. This will help for normalization of user supplied input.

For `executor/simple.go`, it sets the `CharsetDatabase`, and then the `CollationDatabase`. It is safe because setting the `CharsetDatabase` is actually a redundant step, because the set for `CollationDatabase` will overwrite it in its set function.   However, there is an exception to this if the Collation is empty:

https://github.com/pingcap/tidb/blob/6889ab44a5bf21ec0465cd84a4baf68072f1bd3c/executor/simple.go#L564-L567

I think the better fix for this is to change `executor/simple.go` to no longer attempt to set the character set at all, and just try and set the collation\*. The function `getDefaultCollate` will need a small change to return the default collation if the charset specified is invalid\*

\* The commits which addresses these issues are https://github.com/pingcap/tidb/pull/24100/commits/075be93d3de4fd07f5042ab9bf5f448cc7080dcd and https://github.com/pingcap/tidb/pull/24100/commits/9e14457c09bb362c09b51105fc79a1804819fa65

Reply:
An informative explanation.